blog/2024-04-16-custom-components.md

```diff
@@ -0,0 +1,198 @@
+
+For example, say you wanted to use an [image zoom](https://www.npmjs.com/package/react-medium-image-zoom) component that lets you click on an image to zoom in. Rather than figuring out how to wrap it, you can now run
+
+```bash
+pip install reflex-image-zoom
+```
+
+And use it right away like any other Reflex component.
+
+```python demo exec
+import reflex as rx
+from reflex_image_zoom import image_zoom
+
+def zoom():
+    src = "/blog/custom_components.webp"
+    return rx.hstack(
+        rx.vstack(
+            rx.text("Original Image"),
+            rx.image(src=src),
+        ),
+        rx.vstack(
+            rx.text("Click to Zoom Image"),
+            image_zoom(
+                rx.image(src=src),
+                width="50%",
+            )
+        ),
+    )
+```
+
+## High-Level Components
+
+You can also create custom components that don't wrap React, but instead are built out of existing Reflex components. For example, you can define custom navigation bars, modals, or even entire pages as custom components.
+
+In our [intro tutorial]({tutorial.final_app.path}) we share the code for creating a chat component in Reflex. Having the full code gives you full flexibility in how your chat component works and appears, but sometimes you just want to drop in a chat component and not worry about the details.
+
+With custom components, we now have a [reflex-chat](https://github.com/picklelo/reflex-chat/) package that you can install with `pip` and use in your app.
+
+```bash
+pip install reflex-chat
+```
+
+You can then import the chat component into your app and use it like any other Reflex component. All the styling and logic is encapsulated, you only need to pass in the actual LLM logic.
+
+```python exec
+import reflex as rx
+from openai import OpenAI
+from reflex_chat import chat
+
+try:
+    client = OpenAI()
+except:
+    client = None
+
+# Only define your logic, the chat component handles the rest.
+async def run_llm(chat_state):
+    # Start a new session to answer the question.
+    print(chat_state.messages)
+    session = client.chat.completions.create(
+        model="gpt-3.5-turbo",
+        messages=chat_state.messages,
+        stream=True,
+    )
+
+    # Stream the results, yielding after every word.
+    for item in session:
+        if hasattr(item.choices[0].delta, "content"):
+            answer_text = item.choices[0].delta.content
+            # Ensure answer_text is not None before concatenation
+            chat_state.append_to_response(answer_text)
+            yield
+```
+
+```python demo
+rx.box(
+    chat(process=run_llm),
+    height="500px",
+    width="100%"
+)
+```
+
+```md alert info
+# Just set `process` prop of the chat component to connect to your LLM.
```

Comment:
the padding on this box looks a little off; not sure if there's something we can do about that tho

Reply:
Will look in a follow up